### ISSUE TITLE
Thread and Thread pool metric automatic instrumentation

### ISSUE THREAD

**Issue by Falmarri** [open-telemetry/opentelemetry-java-instrumentation#11149]
### Is your feature request related to a problem? Please describe.

Getting thread metrics is pretty important, and is supported by most APMs. Before I can get all of my service owners to switch off of whatever APM they're currently using, I need at least a reasonable degree of feature parity in the metrics.

### Describe the solution you'd like

Thread metrics available to export to otel and/or serve via prometheus. Including 

- number of current threads total
- number of threads used out of a thread pool
- number of runnables executed by that pool
- pool queue size

**Comment by pepeshore** [comment]
I have implemented the collection of thread pool metrics in my forked repository.
including the following metrics:
- thread_pool_core_pool_size
- thread_pool_max_pool_size
- thread_pool_active_thread_count
- thread_pool_current_thread_count
- thread_pool_max_thread_count
- thread_pool_scheduled_task_count
- thread_pool_completed_task_count
- thread_pool_rejected_task_count
- thread_pool_queue_size
labels:
- thread.name.pattern(e.g. http-nio-8080-*)
- thread.pool.class.name(e.g. java.util.concurrent.ThreadPoolExecutor)

if this is helpful, i'd like to create a pr

**Comment by AlchemyDing** [comment]
Refer to [micrometer](https://docs.micrometer.io/micrometer/reference/reference/jvm.html)

**Comment by AlchemyDing** [comment]
I think this is necessary, but the name of metrics needs to be considered.
cc @trask @laurit @steverao

**Comment by steverao** [comment]
It looks like reasonable feature, @open-telemetry/java-instrumentation-approvers WDYT?

**Comment by laurit** [comment]
I think it is a reasonable request, but it is too vague. It would be helpful to know exactly which thread pools the users are interested it. It should already be possible to monitor many thread pools using https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/instrumentation/jmx-metrics/javaagent though users may need to write the configuration themself.

**Comment by pepeshore** [comment]
> I think it is a reasonable request, but it is too vague. It would be helpful to know exactly which thread pools the users are interested it. It should already be possible to monitor many thread pools using https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/instrumentation/jmx-metrics/javaagent though users may need to write the configuration themself.

Most user-defined thread pools cannot be monitored through JMX metrics. And why do you think it si too vague,  you don't know which thread pool should be monitored?

**Comment by laurit** [comment]
> And why do you think it si too vague, you don't know which thread pool should be monitored?

No, I don't. There are many ways one could implement a thread pool, there won't be a catch all solution. Even something seemingly simple, like instrumenting `ThreadPoolExecutor`, immediately gets complicated when you consider that you somehow need to identify which `ThreadPoolExecutor` the metrics belong to.

**Comment by pepeshore** [comment]
> No, I don't. There are many ways one could implement a thread pool

Indeed, there are many implementations of thread pools. However, it is sufficient to monitor the mainstream thread pools. Similar to the existing monitoring of connection pools, there are also various implementations of connection pools, but we only monitor a part of them, such as c3p0, dbcp, and so on.

**Comment by pepeshore** [comment]
>  Even something seemingly simple, like instrumenting `ThreadPoolExecutor`, immediately gets complicated when you consider that you somehow need to identify which `ThreadPoolExecutor` the metrics belong to.

I considered this issue as well. When I added two dimensions to the metrics related to thread pools, namely thread.name.pattern (e.g., http-nio-8080-*) and thread.pool.create.stack (indicating the method stack that created this thread pool), the association between a metric and its thread pool became clear.

**Comment by pepeshore** [comment]
@laurit Hi laurit，What do you think about my opinion？I do think is't a useful feature

**Comment by laurit** [comment]
@pepeshore sure, a PR would be welcome

**Comment by zhangjiabin1010** [comment]
Have the same requirements.Is there any progress ?

**Comment by akats7** [comment]
Hi all, are cpu time thread metrics in scope for this, I can also open a new issue.

**Comment by ixiaoyi93** [comment]
Monitoring custom thread pools is a common requirement, I hope someone will continue to follow up.

**Comment by YaoYingLong** [comment]
@trask @laurit I have opened #19277 as an initial implementation for this issue.

The PR adds automatic metrics for JDK `ThreadPoolExecutor` and JDK 21 `ThreadPerTaskExecutor`. Its purpose is to make the semantic-convention discussion concrete, not to settle the final metric model prematurely.

## Current implementation

`ThreadPoolExecutor` is pre-registered at construction time without creating metrics or invoking its `ThreadFactory`. Metrics begin only when the first real worker thread starts, so executor identity is based on an actual application worker thread and no synthetic thread is created.

If `setThreadFactory()` changes the factory, metrics are updated when the next worker from that factory starts. Generic JDK instrumentation uses `jvm.executor.owner.name=unknown`; future framework instrumentation can call `reregister(...)` to supply an owner name and `all` or `trailing` thread-name normalization. This can happen before initial registration or after metrics are active.

Metrics are removed only after the executor has actually entered shutdown. Weak references in the registration and callbacks ensure that an unclosed executor is not retained solely by metrics instrumentation.

## Candidate metric definitions

| Metric | Instrument | Unit | Attributes |
| --- | --- | --- | --- |
| `jvm.executor.thread.count` | Observable UpDownCounter | `{thread}` | `jvm.executor.name`, `jvm.executor.owner.name`, `jvm.executor.type`, `jvm.executor.state` |
| `jvm.executor.thread.core` | Observable UpDownCounter | `{thread}` | `jvm.executor.name`, `jvm.executor.owner.name`, `jvm.executor.type` |
| `jvm.executor.thread.max` | Observable UpDownCounter | `{thread}` | `jvm.executor.name`, `jvm.executor.owner.name`, `jvm.executor.type` |
| `jvm.executor.queue.size` | Observable UpDownCounter | `{task}` | `jvm.executor.name`, `jvm.executor.owner.name`, `jvm.executor.type` |
| `jvm.executor.queue.remaining` | Observable UpDownCounter | `{task}` | `jvm.executor.name`, `jvm.executor.owner.name`, `jvm.executor.type` |
| `jvm.executor.task.completed` | Observable Counter | `{task}` | `jvm.executor.name`, `jvm.executor.owner.name`, `jvm.executor.type` |
| `jvm.executor.task.rejected` | Counter | `{task}` | `jvm.executor.name`, `jvm.executor.owner.name`, `jvm.executor.type` |

`ThreadPerTaskExecutor` currently reports only active thread count. `ScheduledThreadPoolExecutor` and queue-wait duration are not included.

## Candidate attribute definitions

- `jvm.executor.name` is derived from the first actual worker thread name. `all` normalizes `tomcat-8080-exec-1` to `tomcat-*-exec-*`; `trailing` normalizes it to `tomcat-8080-exec-*`.
- `jvm.executor.owner.name` identifies the consuming component. Generic JDK instrumentation uses `unknown`; framework instrumentation can re-register an executor with values such as `tomcat` or `jetty`.
- `jvm.executor.type` is the fully qualified executor implementation class name.
- `jvm.executor.state` has `active` and `idle` values for `jvm.executor.thread.count`.

I propose moving this forward in stages:

1. Define experimental JVM executor metric semantic conventions in open-telemetry/semantic-conventions.
2. Evolve the reusable registration path into an incubating shared helper after those semantics are agreed.
3. Start with standard JDK executor implementations.
4. Reuse the helper from framework-specific instrumentations later.

The main open questions are:

1. Should the namespace be `jvm.executor.*` or `jvm.thread_pool.*`? I prefer `jvm.executor.*`, because Java's standard abstraction is `Executor`.
2. Should `jvm.executor.owner.name`, `jvm.executor.type`, and `jvm.executor.state` be standardized attributes?
3. Are the proposed `all` and `trailing` name-normalization rules appropriate for framework thread names?
4. Should scheduled-executor metrics and queue-wait duration be part of the first semantic-convention scope or follow separately?
